gooddata-pandas/gooddata_pandas/result_convertor.py

```diff
@@ -0,0 +1,315 @@
+    """per-dimension grand total headers"""
+
+
+@dataclass
+class _AccumulatedData:
+    """
+    Utility class to offload code from the function that extracts all data and headers for a
+    particular paged result. The method drives the paging and calls out to this class to accumulate
+    the essential data and headers from the page.
+    """
+
+    data: Any = field(init=False)
+    data_headers: List[Optional[_DataHeaders]] = field(init=False)
+    grand_totals: List[Optional[List[_DataArray]]] = field(init=False)
+    grand_totals_headers: List[Optional[_DataHeaders]] = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.data = []
+        self.data_headers = [None, None]
+        self.grand_totals = [None, None]
+        self.grand_totals_headers = [None, None]
+
+    def accumulate_data(self, from_result: ExecutionResult) -> None:
+        """
+        if one-dimensional result, the data is single array, so this adds the elements of that array into 'data'
+        if two-dimensional, the data is array of arrays, so this adds as many arrays as there are table rows
+        """
+        self.data.extend(from_result.data)
+
+    def extend_existing_row_data(self, from_result: ExecutionResult) -> None:
+        offset = from_result.paging_offset[0]
+
+        for i in range(len(from_result.data)):
+            self.data[offset + i].extend(from_result.data[i])
+
+    def accumulate_headers(self, from_result: ExecutionResult, from_dim: int) -> None:
+        """
+        Accumulates headers for particular dimension of a result into the provided `data_headers` array at the index
+        matching the dimension index. This will mutate the `data_headers`
+        """
+        if self.data_headers[from_dim] is None:
+            self.data_headers[from_dim] = from_result.get_all_headers(dim=from_dim)
+        else:
+            for idx, headers in enumerate(from_result.get_all_headers(dim=from_dim)):
+                cast(_DataHeaders, self.data_headers[from_dim])[idx].extend(headers)
+
+    def accumulate_grand_totals(
+        self, from_result: ExecutionResult, paging_dim: int, response: BareExecutionResponse
+    ) -> None:
+        """
+        accumulates grand totals from the results; processes all grand totals on all dimensions; the method
+        needs to know in which direction is the paging happening so that it can append new grand total data.
+        """
+        grand_totals = from_result.grand_totals
+        if not len(grand_totals):
+            return
+
+        # get dimension indexes mapping from response like {"dim1": 0, "dim0": 1}
+        dim_idx_dict = {dim["localIdentifier"]: idx for idx, dim in enumerate(response.dimensions)}
+
+        for grand_total in grand_totals:
+            # 2-dim results have always 1-dim grand totals (3-dim results have 2-dim gt but DataFrame stores 2D only)
+            dims = grand_total["totalDimensions"]
+            assert len(dims) == 1
```

Comment:
Please, add a message after assert explaining the assert.

assert len(dims) == 1, "What does this assert mean."

Reply:
fixed in last commit